packages/dynamite/dynamite/lib/src/helpers/version_checker.dart

```diff
@@ -0,0 +1,101 @@
+  'built_collection': Version.parse('5.0.0'),
+  'built_value': Version.parse('8.0.0'),
+  'collection': Version.parse('1.0.0'),
+  'dynamite_runtime': Version.parse('0.0.1'),
+  'meta': Version.parse('1.0.0'),
+  'universal_io': Version.parse('2.0.0'),
+  'uri': Version.parse('1.0.0'),
+};
+
+final devDependencies = {
+  'built_value_generator': Version.parse('8.8.0'),
+};
+
+/// Checks whether the correct version of `dynamite_runtime` is present in the
+/// pubspec.yaml file.
+Future<({bool hasFatal, String messages})> helperVersionCheck(final BuildStep buildStep) async {
+  final pubspecAsset = AssetId(buildStep.inputId.package, 'pubspec.yaml');
+
+  if (!await buildStep.canRead(pubspecAsset)) {
+    throw Exception(
+      'Failed to read the pubspec.yaml file. '
+      'Version constraints of required packages can not be validated.',
+    );
+  }
+
+  final pubspecContent = await buildStep.readAsString(pubspecAsset);
+  final pubspec = Pubspec.parse(pubspecContent, sourceUrl: pubspecAsset.uri);
+
+  final messages = StringBuffer();
+  var hasFatal = false;
+
+  for (final constraint in dependencies.entries) {
+    final result = _validateVersion(pubspec.dependencies, constraint.key, constraint.value);
+
+    if (result.message != null) {
+      messages.writeln(result.message);
+    }
+
+    if (result.isFatal) {
+      hasFatal = true;
+    }
+  }
+
+  for (final constraint in devDependencies.entries) {
+    final result = _validateVersion(pubspec.devDependencies, constraint.key, constraint.value);
+
+    if (result.message != null) {
+      messages.writeln(result.message);
+    }
+
+    if (result.isFatal) {
+      hasFatal = true;
+    }
+  }
+
+  return (hasFatal: hasFatal, messages: messages.toString());
+}
+
+({bool isFatal, String? message}) _validateVersion(
+  final Map<String, Dependency> dependencies,
+  final String packageName,
+  final Version minVersion,
+) {
+  final dependency = dependencies[packageName];
+
+  if (dependency == null) {
+    return (
+      isFatal: false,
+      message: 'Could not find the required dependency on `$packageName` in pubspec.yaml file.'
+          ' Compatibility check is being skipped.',
+    );
+  } else if (dependency is! HostedDependency) {
+    return (
+      isFatal: false,
+      message: 'The dependency on `$packageName` is not a `$HostedDependency`. '
+          'Compatibility check is being skipped.',
+    );
+  } else {
+    final constraint = dependency.version;
+    final constraintTooLowMessage = 'The version constraint "$constraint" on `$packageName` allows '
+        'versions before $minVersion which is not allowed.';
+
+    if (constraint is Version && constraint < minVersion) {
+      return (isFatal: true, message: constraintTooLowMessage);
+    }
+
+    final range = constraint as VersionRange;
+    final rangeMin = range.min;
+
+    if (rangeMin == null || rangeMin < minVersion) {
+      return (isFatal: true, message: constraintTooLowMessage);
+    }
```

Comment:
We also need to limit the upper version since a new major version can/will have breaking changes. The maximum exclusive version could be calculated on the fly as `${major+1}.0.0`.